### ISSUE TITLE
Bad System Call (core dumped)

### ISSUE THREAD

**Issue by abbasmaheryar** [stoeckmann/xwallpaper#28]
OS: Arch Linux
WM: dwm

Hello.

I'm not exactly familiar with providing the necessary debugging information to assist you all in helping me with this problem, but I'll provide what I can. 

Yesterday when I booted up my computer, my wallpaper script ran normally upon startx, and I could see my wallpaper. Today when I booted my computer this morning, my script failed, and I could not see my background. When I ran the script manually, xwallpaper returned this:
```
[abbas@archchad ~]$ xwallpaper --zoom .local/share/prophecy_jump_puzzle.jpg
zsh: invalid system call (core dumped)  xwallpaper --zoom .local/share/prophecy_jump_puzzle.jpg
[abbas@archchad ~]$
```
Any help would be appreciated. Pardon me if I'm not providing enough information; I'm just not sure where to look to analyze this problem. Thanks

**Comment by abbasmaheryar** [comment]
Something else that might be useful: 
```
[abbas@archchad ~]$ xwallpaper --version
zsh: invalid system call (core dumped)  xwallpaper --version
[abbas@archchad ~]$
```
I had this problem when I tried the official arch linux package and from building from the git tree.

**Comment by collidedscope** [comment]
Just chiming in to confirm that the issue exists for at least `spectrwm` as well, so it's possibly WM-agnostic. `xwallpaper` hasn't changed recently, so the issue is obviously further upstream. The core dump for `xwallpaper --version` indicates that we're being killed merely for trying to call `puts()`:

```
Stack trace of thread 15540:
#0  0x00007fecea57567e fstatat (libc.so.6 + 0xef67e)
#1  0x00007fecea4fa763 _IO_file_doallocate (libc.so.6 + 0x74763)
#2  0x00007fecea509090 _IO_doallocbuf (libc.so.6 + 0x83090)
#3  0x00007fecea508228 _IO_file_overflow@@GLIBC_2.2.5 (libc.so.6 + 0x82228)
#4  0x00007fecea5072d6 _IO_file_xsputn@@GLIBC_2.2.5 (libc.so.6 + 0x812d6)
#5  0x00007fecea4fcde8 _IO_puts (libc.so.6 + 0x76de8)
#6  0x000055637d0b6bae n/a (/home/collided/xwallpaper/xwallpaper + 0x5bae)
```

Moreover, `strace` indicates the invalid system call is to `newfstatat()`:

```
prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)  = 0
seccomp(SECCOMP_SET_MODE_FILTER, 0, {len=93, filter=0x56439ab81480}) = 0
newfstatat(1, "",  <unfinished ...>)    = ?
+++ killed by SIGSYS (core dumped) +++
zsh: invalid system call (core dumped)  strace ./xwallpaper --version
```

The following notice from `man 2 stat` seems pertinent:

> No permissions are required on the file itself, but—in the case of **stat**(), **fstatat**(), and **lstat**()—execute (search) permission is required on all of the directories in `pathname` that lead to the file.

Given that the syscall dies right after trying to process the empty `pathname`, it stands to reason that it should be something like `/dev/fd/1` for standard output, but I can't say whether that's true or how it would come to be the case. Given that `xwallpaper` is "different" in that it strives to be secure with its various sandboxing features, I'm inclined to think the root cause lies somewhere in that department, but otherwise have reached the limit of my ability to provide insight into the matter.

I haven't yet bothered to confirm this suspicion, but I suspect some recent change to the Linux kernel (in particular, something to do with `seccomp`) is the culprit and that downgrading would suffice as a temporary fix, but I'd like to be of any assistance I can in determining how to avoid that course of action.

**Comment by frei-0xff** [comment]
I have the same issue. It happened after glibc upgraded from 2.32-5 to 2.33-3
Downgrading back to 2.32-5 temporally fixes the issue.

**Comment by Swy7ch** [comment]
Can confirm this: the issue emerged after upgrading glibc, and is "fixed" by downgrading it. Should we report it to GNU org, then ?

**Comment by collidedscope** [comment]
I believe I've diagnosed the issue. glibc is certainly part of it, as [the 2.33 release](https://github.com/bminor/glibc/commit/9826b03b747b841f5fc6de2054bf1ef3f5c4bdf3) seems to've involved quite a few under-the-hood changes surrounding the `fstatat()` syscall. In addition to that, it seems I was in the ballpark in suggesting that it might have something to do with `xwallpaper`'s security-oriented approach, specifically `seccomp`:

```c
seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(newfstatat), 0) ||
```

Adding the above line to `seccomp.c` and rebuilding suffices to clear up the core dump on my Arch machine, but I've refrained from submitting a patch since I'm not sure it's a sufficiently general solution.

**Comment by stoeckmann** [comment]
Hi @collidedscope and everyone else who helped to diagnose this early on.

Do not hesitate to submit patches. The worst thing that could happen is that it's discussed and rejected. :)
I'll integrate this, after all the system call should not be able to be abused for harmful actions in case of an attack based on image files. I follow a white-list approach compared to a black-list approach here, so the analysis is correct.

It is also very helpful if xwallpaper is used on more architectures. When I tested it on a rasperry pi system with an older version, I also spotted some more system calls used by libjpeg internally. If there's a better way to specify groups of system calls like OpenBSD's pledge does, I would like to hear about it!

**Comment by stoeckmann** [comment]
Considering the positive feedback @collidedscope's reply got, I assume that more people have verified that this single line is sufficient for more setups.

If this is confirmed by building the current sources, then I will release a new version tomorrow (Sunday).

**Comment by Swy7ch** [comment]
I just built the latest version (including commit `90d906b`) and it does work on Arch Linux, running dwm as window manager.

**Comment by collidedscope** [comment]
And of course I can vouch for Arch + `spectrwm`, but it's not radically different from `dwm` (having been directly inspired by it), so we don't have an especially broad sample. Still, I'm inclined to think that folks using `xwallpaper`‒a tool that does one thing *remarkably* well‒are of a predisposition to running the sort of environment represented in this thread (namely Arch), where the one-line fix certainly seems to squash the bug.

@stoeckmann, I could try a handful of other configurations if you'd like to make absolutely sure the patch has broader coverage, but your utility has hitherto been totally rock-solid for me (to the point that I forget it's there), so I reckon this was mostly just a case of bad luck.

**Comment by stoeckmann** [comment]
I have released version 0.6.6 which contains the commit.

Thanks everyone for reporting, investigating and actually fixing it! :)

Considering the idea of "release early, release often" I see no need to check it on even more systems, @collidedscope. This is a rat race anyway, since we have no control of all involved libraries. There are not much, but theoretically even an optimization setting in compilers can change things. Not an optimization flag, but I had such effects in man-db's seccomp usage when I cross-compiled it as well as glibc. They were unable to get all information in configure run and therefore included system calls which were not expected by the sandbox.

We could argue that it's a distribution topic anyway, because they bundle software packages and various versions. But that's not feasable either from a realistic perspective. It would be way too much work for them. So it's not optimal but "it works." Most of the time. ;)

I close this issue, please reopen if it happens again.

@abbasmaheryar: The report was sufficient. If I know the architecture and distribution, I should be able to reproduce it. Running xwallpaper with strace and posting the last lines definitely helps as well. Please open a new issue if something like that happens again.
